src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,489 @@
+void Instance::InvokeCommand(HandlerContext & handlerContext)
+{
+    switch (handlerContext.mRequestPath.mCommandId)
+    {
+    case ModeBase::Commands::ChangeToMode::Id:
+        ChipLogDetail(Zcl, "ModeBase: Entering handling ChangeToModeWithStatus");
+
+        HandleCommand<Commands::ChangeToMode::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & commandData) { HandleChangeToMode(ctx, commandData); });
+    }
+}
+
+// List the commands supported by this instance.
+CHIP_ERROR Instance::EnumerateAcceptedCommands(const ConcreteClusterPath & cluster,
+                                               CommandHandlerInterface::CommandIdCallback callback, void * context)
+{
+    callback(ModeBase::Commands::ChangeToMode::Id, context);
+    return CHIP_NO_ERROR;
+}
+
+// List the commands generated by this instance.
+CHIP_ERROR Instance::EnumerateGeneratedCommands(const ConcreteClusterPath & cluster, CommandIdCallback callback, void * context)
+{
+    callback(ModeBase::Commands::ChangeToModeResponse::Id, context);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::EncodeSupportedModes(const AttributeValueEncoder::ListEncodeHelper & encoder)
+{
+    for (uint8_t i = 0; true; i++)
+    {
+        ModeOptionStructType mode;
+
+        // Get the mode label
+        char buffer[kMaxModeLabelSize];
+        MutableCharSpan label(buffer);
+        auto err = mDelegate->GetModeLabelByIndex(i, label);
+        if (err == CHIP_ERROR_PROVIDER_LIST_EXHAUSTED)
+        {
+            return CHIP_NO_ERROR;
+        }
+        ReturnErrorOnFailure(err);
+
+        mode.label = label;
+
+        // Get the mode value
+        ReturnErrorOnFailure(mDelegate->GetModeValueByIndex(i, mode.mode));
+
+        // Get the mode tags
+        ModeTagStructType tagsBuffer[kMaxNumOfModeTags];
+        DataModel::List<ModeTagStructType> tags(tagsBuffer);
+        ReturnErrorOnFailure(mDelegate->GetModeTagsByIndex(i, tags));
+        mode.modeTags = tags;
+
+        ReturnErrorOnFailure(encoder.Encode(mode));
+    }
+    return CHIP_NO_ERROR;
+}
+
+// Implements the read functionality for complex attributes.
+CHIP_ERROR Instance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    switch (aPath.mAttributeId)
+    {
+    case Attributes::CurrentMode::Id:
+        ReturnErrorOnFailure(aEncoder.Encode(mCurrentMode));
+        break;
+    case Attributes::StartUpMode::Id:
+        ReturnErrorOnFailure(aEncoder.Encode(mStartUpMode));
+        break;
+    case Attributes::OnMode::Id:
+        ReturnErrorOnFailure(aEncoder.Encode(mOnMode));
+        break;
+    case Attributes::SupportedModes::Id:
+        Instance * d   = this;
+        CHIP_ERROR err = aEncoder.EncodeList([d](const auto & encoder) -> CHIP_ERROR { return d->EncodeSupportedModes(encoder); });
+        return err;
+    }
+    return CHIP_NO_ERROR;
```

Comment:
This needs to actually have a case for FeatureMap, no?

Reply:
Yeah. Missed that. Added [here](https://github.com/project-chip/connectedhomeip/pull/27504/commits/521b61fc6c7bc06d7db0f6bd7f265b00a9b72dad).